### PR TITLE
Update gitignore

### PR DESCRIPTION
**Pull request by Sollace**
Added all the eclipse files and generated binaries to the gitignore. This was part of a previous PR, but i figured it would be better merged on its own.

Edit: Exhibit A
image

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,12 @@
 # Ignore everything
 /*
 
+# Ignore eclipse files
```
Comment (modmuss50) [inline .gitignore]: Why is this needed? we use an gitinclude here? (see first line).

If this is for subproject stuff see how its done for the other files such as .gradle bellow

Comment (Sollace) [reply .gitignore]: I don't know what gitinclude is, but every time I work on the project I have hundreds of unchecked files from the sub-modules because they're not excluded.

Comment (Sollace) [reply .gitignore]: Added an image to illustrate what this PR fixes.

Comment (modmuss50) [reply .gitignore]: This isnt the correct fix, look at the subproject section that already exists.

Comment (Sollace) [reply .gitignore]: I see this:

```
# Subprojects
!/fabric-*
/fabric-*/.gradle
/fabric-*/build
```

But that doesn't correctly exclude all your files. If you're going to do that you have to do:
```
# Subprojects
!/fabric-*/src
!/fabric-*/build.gradle
/fabric-*/.gradle
/fabric-*/build
```

Comment (Sollace) [reply .gitignore]: Actually, no. That still doesn't include the `src` folder.

Comment (Sollace) [reply .gitignore]: You should just exclude the files you don't want appearing in the repo. Including them individually is exactly what leads to these kinds of problems.

**Comment by grondag** [comment]
I can confirm the .gitignore files as is don't work in my environment.  They don't properly exclude eclipse varia and even some Idea files show up.

I am baffled as to why - would think the original fabric contributors would have the same problem.  I worked around the problem with a .gitignore_global file but that's not a good a solution for new contributors in general.

The "gitinclude" approach (exclude everything, then include what you want) seems like it _should_ work.  Can it be salvaged?

**Comment by sfPlayer1** [comment]
IMO the whitelist approach is quite a mess and unreliable, controlling the artefacts isn't particularly difficult.

The exclude pattern may not work because * usually doesn't cover dot files.

**Comment by Sollace** [comment]
Can we get a consensus on what's going to happen with this?

As it stands, the way fabric has it's `.gitignore` _does not work_, and the way devs are using it is _objectively wrong_ (in my subjective opinion)

**Comment by asiekierka** [comment]
>objectively wrong (in my subjective opinion)

**Comment by grondag** [comment]
I discussed this briefly with Player, Asie and Modmuss.  We support changing the .gitignore to use the standard ignore approach vs. exclude all then include.

Suggest you simply redo the whole thing in the customary way and then verify it works with both Idea and Eclipse.   Should be acceptable at that point.

**Comment by Sollace** [comment]
Sure. I may look at adapting one I use for other projects.

File: .gitignore [hunk 2]
```diff
@@ -1,25 +1,32 @@
...
-!/HEADER
-!/Jenkinsfile
-!/LICENSE
-!/README.md
-!/settings.gradle
+# Build artifacts
+bin/
+build/
+jars/
+
```
Comment (grondag) [inline .gitignore]: Tested locally on current Fabric project and needed to add `out/`

Otherwise perfect - eclipse wise.

**Comment by grondag** [comment]
@modmuss50 Could you have another look and confirm this works for Idea? Would be awfully nice to have it for us Eclipse users.  I always need to unbork my local gitignore after every reset.

**Comment by Sollace** [comment]
@grondag Idea should be working fine without problems. I specifically set it up so I could test the gitignore. That said, there might still be some things I missed like the `out/` folder.

**Comment by grondag** [comment]
I assumed as much. Always best to get independent verification when practical.

**Comment by Sollace** [comment]
@sfPlayer1 Every module is going to have their own junk generated in them, so I wouldn't recommend doing that.

**Comment by modmuss50** [comment]
Merged in manually here: https://github.com/FabricMC/fabric/compare/9b03381f03f4...9f55aa704f28

Thanks.
